_posts/2023-02-01-opentelemetry.md

```diff
@@ -0,0 +1,214 @@
+        - name: OTEL_EXPORTER_OTLP_ENDPOINT
+          value: http://simplest-collector:4317
+  tracing:
+    type: opentelemetry
+```
+
+With the HTTP bridge up and running, we can use any HTTP client for sending and receiving messages.
+It could be a simple command line tool like `cURL`, a more UI oriented like [Postman](https://www.postman.com/) or whatever is your preferred one.
+You could also use an application you developed with your preferred language.
+For the purpose of this blog post, I used Postman by having a producer sending a couple of messages and the consumer getting them.
+
+When the HTTP clients exchange the messages we can see the traces and the corresponding spans in the Jaeger Web UI.
+
+![](/assets/images/posts/2023-02-01-opentelemetry-01.png)
+
+In the above picture, the HTTP producer sent two messages and they are received by the HTTP consumer.
+
+When the HTTP client sends a producer request, the bridge generates two spans linked together as part of the same trace:
+
+* the first span defines the HTTP `send` operation requested by the client. This span is created by the bridge application itself instrumented with OpenTelemetry.
+* the second span defines the low level Kafka `send` on the `my-topic` topic which is a "CHILD_OF" the first one. This span is created by the instrumented Kafka client provided by the OpenTelemetry project. It uses a producer interceptor for creating the span.
+
+![](/assets/images/posts/2023-02-01-opentelemetry-02.png)
+
+In the same way, when the HTTP client send a poll request for receiving the messages, the bridge generates three spans linked together as part of the same trace:
+
+* the first span defines the HTTP `poll` operation requested by the client. This span is created by the bridge application itself instrumented with OpenTelemetry.
+* the second and third spans define the low level Kafka `receive` on the `my-topic` topic related to the messages received. They are both "CHILD_OF" the first span but at the same time they are referencing the external trace related to the producer with a "FOLLOW_FROM" relationship. These spans are created by the instrumented Kafka client provided by the OpenTelemetry project. It uses a consumer interceptor for creating them.
+
+![](/assets/images/posts/2023-02-01-opentelemetry-03.png)
+
+It is interesting to highlight that, due to the asynchronous nature of the communication between producer and consumer through the HTTP bridge and Kafka, the producing and consuming parts are traced with two different traces.
+The consuming related spans are linked to the producing trace spans by using a "FOLLOW_FROM" relationship.
+This takes into account that same messages could be consumed by multiple consumers or read again by the same consumer in the future.
+Having different consumer traces linked to the producer one makes more sense than having an heavy and not so easy to navigate bigger trace.
+
+## Kafka Connect, Mirror Maker(s) and clients
```

Comment:
I guess it would be too complicated for an example. But I think it would be great to be able to show here some longer chain -> e.g. Kafka producer, streams API, consumer, Connect with connector etc. as a screenshot from the Jaeger UI which would show how it might look like when used in all the components.

Reply:
I decided to not go in details here on purpose. If we do, we are going to have the same blog post you wrote long time ago with what you are asking and I don't see much value in it. Between now and then what's changing is the configuration of the custom resources and the env vars to use. The Jaeger UI will look like mostly the same.
So I decided to go with a simple example with the HTTP bridge which was not covered in your blog post.

Reply:
It is ultimately your decision. But I think the blog post is right now is very focused on existing users of OpenTracing. If you are new user who does not use tracing yet, you will not take away much from it -> you will likely not gain any understanding on what tracing is and why it could be useful. For example showing screenshots with proper flow with clients in the previous section or some nice screeenshot with multiple different Kafka applications here would IMHO go long way to give something also to the new users to make them understand why it might be useful.

Reply:
Maybe it would make sense to split it into two blog posts? One as a new user of tracing with more details about tracing itself with a little bit more complex use-case and one (this one) as a guide for changes related to tracing and 0.32?

Reply:
Tbh I think two blog posts are too much for this. I will provide some screenshots for Kafka Connect and Mirror Maker, it should be enough. It's useless to write 3 blog posts in the end on the same subject with one of them about migration to be really short.